### ISSUE TITLE
Conditionally Use New Resource

### ISSUE THREAD

**Issue by maxmanders** [hashicorp/terraform#21775]
### Terraform Version

```
$ terraform -v
Terraform v0.11.14
+ provider.aws v2.0.0

Your version of Terraform is out of date! The latest version
is 0.12.2. You can update by downloading from www.terraform.io/downloads.html
```

I'm also using version 2.0.0 of _terraform-provider-aws_.

I have tried to reproduce the problem with Terraform 0.12.2 and _terraform-provider-aws_ 2.15.0 with no success, so my problem appears to be fixed in later versions. I plan to upgrade to the latest versions of all components, but that's not an option right now.

If you could help me understand the issue, and suggest any workarounds, that would be very much appreciated.

### Terraform Configuration Files

See [this Gist](https://gist.github.com/maxmanders/946a7da7f12dc7ac768aeea1b6443eb3).

The file `tf_0_main.tf` creates the initial state and resources. After which, application of `tf_1_main.tf` exhibits the problematic behaviour.

### Debug Output

See Gist

### Crash Output

N/A

### Expected Behavior

Given a conditional local variable:

```hcl
some_condition = true
chosen_sg = "${local.some_condition ? aws_security_group.sg.id : join(",", var.security_groups)}"
```

I'd expect to be able to use `local.chosen_sg` where a list of security group IDs is expected, to modify an EC2 host to use a different, newly created, security group.

Were I instead to change the behaviour thusly

```hcl
some_condition = false
chosen_sg = "${local.some_condition ? aws_security_group.sg.id : join(",", var.security_groups)}"
```

I'd expect the new security group to be created, without any change to the configuration of the EC2 instance.

### Actual Behavior

However the conditional expression is evaluated, an update to the EC2 instance is triggered.

I note however, that if I wrap the _true_ branch of the ternary expression in quotes, e.g. "aws_security_group.sg.id" and set `some_condition` to `true` then the EC2 instance will try to update the security group list to the string literal "aws_security_group.sg.id". If I set `some_condition` to _false_ then the no-op change I'd expect to see, happens.

It would appear that when trying to use a new resource, i.e. one that hasn't been created, but _would_ be created by the code, in a conditional, the ternary expression is evaluated _oddly_. If I wrap the fully qualified resource reference in quotes, the behaviour is as expected, but only if I don't want to evaluate the ternary expression as _true_.

### Additional Context

In summary, I have existing infrastructure that I want to retain, while introduce a new security group, and optionally using that new security group in place of the existing security group, depending on workspace.

### References

N/A

**Comment by apparentlymart** [comment]
Hi @maxmanders! Sorry for this odd behavior.

I suspect what you are seeing here is Terraform's special unknown value expression behaviors. Whenever an expression result depends on a value which cannot be known until apply time, the result of that expression is also unknown until apply time.

Terraform 0.12 included several changes to improve the precision of those checks so that more expressions can be evaluated without triggering that unknown propagation behavior. In particular, in Terraform 0.12 if the condition expression in a conditional operator produces a known result and the result expression it selects also produces a known result then the known-ness of the other result expression is not considered.

Terraform 0.11's expression evaluator is unfortunately more conservative: it will tend to consider any expression that includes an unknown value as returning unknown, without even trying to process the expression. In your case, I expect the `aws_security_group.sg.id` expression is unknown and therefore Terraform 0.11 is assuming it can't produce a known result from the conditional as a whole, even though we can see it ought to not be considering that possibility at all.

One way to work around this in Terraform 0.11 would be to apply this change in two steps, using `-target`:

```
terraform apply -target=aws_security_group.sg
```

The above will instruct Terraform to only do enough work for that `aws_security_group.sg` resource to be consistent, and to ignore downstream resources. After that succeeds, you can then run `terraform apply` as normal to complete the rest of the updates, using the now-known value of `aws_security_group.sg.id`.

(`-target` is intended only for exceptional circumstances like this where we need to give Terraform an extra hint to work around its conservative behavior. It's not for everyday use as part of a routine Terraform workflow.)

Terraform 0.12's expression evaluator takes a different approach to unknown values that avoids problems like these: rather than detecting unknown values before even starting to evaluate the expression, it will instead evaluate the expressions as much as possible and only produce an unknown result if values encountered during the evaluation are unknown. In the case of a conditional expression, for example, Terraform 0.12 would still return an unknown result if the part before the `?` were unknown, because Terraform then can't predict which of the two values will be selected. But if the condition is known then it will proceed then to evaluate whichever of the result expressions was selected, and disregard the other one.

Since this bug is already fixed in Terraform 0.12, and the fix is impractical to backport (it was "fixed" by replacing the language runtime altogether), I'm going to close this out now. I hope the above is useful to understand the context of the problem and to work around it in the mean time until you are ready to upgrade to Terraform 0.12.

**Comment by maxmanders** [comment]
Thank you so much for your prompt and detailed response Martin! I’m working on upgrading the project in question to $latest as we speak. I’m quite new to Terraform, so your taking the time to give such a thorough reply really helps my learning and understanding!
